[PT Run] Web search should use user's default browser #16549

lncubus opened the pull request:
## Summary of the Pull Request

**What is this about:**
Before this PR PT Run used Application or Default Icon location specified in the registry to find default browser program location. If PT Run was unable to detect program location it used Microsoft Edge.
Also for Opera browser implementation checked internal Opera settings and started a search engine specified with constructed address.

**What is included in the PR:** 
Default Browser Info detection logic updated to use shell command for default browser.
Parameters specified in command are used and search terms are working for Chrome, Firefox, Microsoft Edge and Opera.
If we're unable to find default browser we're using Microsoft Edge.

**How does someone test / validate:** 
Start Power Toys
Go to Power Toys Run
Type ?? to start Web Search plugin
Check your default browser is mentioned in the suggestion area
Close Power Toys
Change your default browser
Repeat previous steps and check that a new default browser shown in the suggestion area

## Quality Checklist

- [x] **Linked issue:** #16549 
- [x] **Communication:** I've discussed this with core contributors in the issue. 
- [ ] **Tests:** Added/updated and all pass
- [ ] **Installer:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Docs:** Added/ updated
- [x] **Binaries:** Any new files are added to WXS / YML
   - [x] No new binaries
   - [ ] [YML for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/pipeline.user.windows.yml#L68) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.

lncubus committed:
Web search does not use my default browser #16549
updated DefaultBrowserInfo logic to get program location from shell open command not icon location

htcfreek commented:
@incubus
What exactly was the problem in the code and what behavior does your change result in?

lncubus commented:
@htcfreek I'd like more information about original problem actually but can only reproduce it with Registry Editor.
However, getting program location with icon definition is not a correct way.
Technically, we can assign any irrelevant icon from any available source to be "default icon" for an application.
I think that was the original issue but have no proof of that now.

htcfreek commented:
@htcfreek 
Can you upload some screenshots of the bug. I can't follow you. Sorry for that.

lncubus commented:
> asking users to try it.
That would help but we need someone with original problem.

jaimecbernardo commented:
> > asking users to try it.
> 
> That would help but we need someone with original problem.

There are some on the issue, right?

lncubus commented:
> @lncubus Can you upload some screenshots of the bug. I can't follow you. Sorry for that.

Browser application registered in HKCR part of registry. The application record have some kind of  icon resource locator and open command command line.

The icon location points to the original application most of the time but it could be different and that would not affect browser functions.

Our original implementation expected icon to point to an application location. It could work most of the time but the default icon could be set to different place and that would not affect default browser running.

New implementation parses command line specified for a default browser.
If that setting is damaged, Windows detects that and asks a user to redefine default browser settings.

Using this approach we will start default browser for the user if the user can start it and that is the expected behavior as far as I can see.

jaimecbernardo reviewed (changes requested):
The value in the registry in https://github.com/microsoft/PowerToys/issues/16549#issuecomment-1079585479
indicates that there's a set way to call the browser with arguments. This PR is just taking the executable path, instead of using the way to call that process in the registry. Is there a way to make this compatible? Does that `%1` get translated into something when called through starting the process?

lncubus committed:
Merge branch 'microsoft:main' into lncubus/web_search_doesnt_use_default_browser

crutkas commented on src/modules/launcher/Wox.Plugin/Common/DefaultBrowserInfo.cs:
```diff
@@ -157,32 +157,42 @@ public static void Update()
+                    string programLocation = GetRegistryValue($@"HKEY_CLASSES_ROOT\{progId}\shell\open\command", null);
+                    if (string.IsNullOrEmpty(programLocation))
+                    {
+                        throw new ArgumentOutOfRangeException(
```
will this crash PowerToys

lncubus replied:
No, this is happening in `try...catch` block.

crutkas commented on src/modules/launcher/Wox.Plugin/Common/DefaultBrowserInfo.cs:
```diff
@@ -157,32 +157,42 @@ public static void Update()
                         Path = string.Equals(FilePath.GetExtension(directProgramLocation), ".exe", StringComparison.Ordinal)
                             ? directProgramLocation
-                            : null;
+                            : throw new ArgumentOutOfRangeException(
```
will this crash PowerToys

lncubus committed:
Refactored DefaultBrowserInfo to start default browser with arguments specified in shell/open/command value for the browser.
Added fallback to Microsoft Edge if no browser information available.
#16549 Web search does not use my default browser

lncubus committed:
Merge branch 'microsoft:main' into lncubus/web_search_doesnt_use_default_browser

lncubus committed:
fixed comment
#16549 Web search does not use my default browser

lncubus commented on src/modules/launcher/Wox.Plugin/Common/DefaultBrowserInfo.cs:
```diff
@@ -157,32 +157,42 @@ public static void Update()
+                    // Get the path to the browser's executable:
+                    // for example "C:\Program Files\SomeCompany\SomeBrowser\Application\launcher.exe" --single-argument %1
+                    string programLocation = GetRegistryValue($@"HKEY_CLASSES_ROOT\{progId}\shell\open\command", null);
+                    if (string.IsNullOrEmpty(programLocation))
```
That's a good point, now we just hope for the MS Edge to be there in the Default Browser Info Update method.

jaimecbernardo commented on src/modules/launcher/Wox.Infrastructure/Helper.cs:
```diff
@@ -148,6 +148,16 @@ public static Process OpenInConsole(string path)
+            if (pattern.Contains("%1", StringComparison.Ordinal))
+            {
+                arguments = pattern.Replace("%1", arguments);
+            }
```
What would happen if the pattern doesn't contain %1? Would it work as well?

lncubus replied:
If there are no %1 we will use `arguments` as is which is the best we can do, I suppose.

htcfreek replied:
This shouldn't happen as %1 is the default place holder for parameters on file associations. If %1 is missing you can't open the file by double click or context menu in explorer.

jaimecbernardo replied:
Got it. THank you!

jaimecbernardo reviewed (commented):
Have a small question, but other than that it LGTM and seems to work well.
Good work!

jaimecbernardo reviewed (approved):
LGTM! Good work!
